File: WoofAndroid/data/repositories/src/main/kotlin/com/perrystreet/woof/repositories/dogs/DogsRepository.kt
package com.perrystreet.woof.repositories.dogs

import com.perrystreet.woof.datasource.dogs.IDogsDataSource
import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.models.dog.DogProfile
import com.perrystreet.woof.models.dog.DogsFeed
import com.perrystreet.woof.models.dog.DogsPage
import com.perrystreet.woof.repositories.dogs.mapper.DogProfileDTOToDomainMapper
import com.perrystreet.woof.repositories.dogs.mapper.DogsPageDTOToDomainMapper
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject

@org.koin.core.annotation.Single
class DogsRepository(
    private val dataSource: IDogsDataSource,
    private val pageMapper: DogsPageDTOToDomainMapper,
    private val profileMapper: DogProfileDTOToDomainMapper,
) {
    private val feed: BehaviorSubject<DogsFeed> = BehaviorSubject.createDefault(DogsFeed.Empty)
    private val profiles: BehaviorSubject<Map<Long, DogProfile>> = BehaviorSubject.createDefault(emptyMap())

    val dogsFeed: Observable<DogsFeed> = feed

    fun getDogs(offset: Int, limit: Int): Single<DogsPage> =
        dataSource.getDogs(offset, limit)
            .map { pageMapper(it) }
            .doOnSuccess { page -> feed.onNext(feed.value!!.append(page)) }

    fun getDogProfile(dog: Dog): Single<DogProfile> =
        profiles.value!![dog.id]?.let { Single.just(it) }
            ?: dataSource.getDogProfile(dog.id)
                .map { profileMapper(it) }
                .doOnSuccess { profile -> profiles.onNext(profiles.value!! + (dog.id to profile)) }

    private fun DogsFeed.append(page: DogsPage): DogsFeed {
        val knownIds = dogs.map { it.id }.toSet()
        val newDogs = page.dogs.filterNot { it.id in knownIds }
        return DogsFeed(dogs = dogs + newDogs, total = page.total)
    }
}

File: WoofIOS/SwiftPackages/Data/Repositories/Sources/Repositories/Dogs/DogsRepository.swift
import Combine
import DI
import DataSource
import Models

@Single
public final class DogsRepository {
    private let dataSource: DogsDataSourceImplementing
    private let pageMapper: DogsPageDTOToDomainMapper
    private let profileMapper: DogProfileDTOToDomainMapper

    private let feed = CurrentValueSubject<DogsFeed, Never>(.empty)
    private let profiles = CurrentValueSubject<[Int: DogProfile], Never>([:])

    public var dogsFeed: AnyPublisher<DogsFeed, Never> {
        feed.eraseToAnyPublisher()
    }

    public func getDogs(offset: Int, limit: Int) -> AnyPublisher<DogsPage, DogsError> {
        dataSource.getDogs(offset: offset, limit: limit)
            .map { [pageMapper] in pageMapper($0) }
            .mapError { _ in DogsError.unavailable }
            .handleEvents(receiveOutput: { [feed] page in feed.send(feed.value.append(page)) })
            .eraseToAnyPublisher()
    }

    public func getDogProfile(dog: Dog) -> AnyPublisher<DogProfile, DogsError> {
        if let profile = profiles.value[dog.id] {
            return Just(profile).setFailureType(to: DogsError.self).eraseToAnyPublisher()
        }
        return dataSource.getDogProfile(dogId: dog.id)
            .map { [profileMapper] in profileMapper($0) }
            .mapError { _ in DogsError.unavailable }
            .handleEvents(receiveOutput: { [profiles] profile in
                var cached = profiles.value
                cached[dog.id] = profile
                profiles.send(cached)
            })
            .eraseToAnyPublisher()
    }
}

private extension DogsFeed {
    func append(_ page: DogsPage) -> DogsFeed {
        let knownIds = Set(dogs.map(\.id))
        let newDogs = page.dogs.filter { !knownIds.contains($0.id) }
        return DogsFeed(dogs: dogs + newDogs, total: page.total)
    }
}
